File: app/backend/sell/pricing.py
"""MRP-inclusive tax arithmetic for a bill line - pure functions, no ORM.

A price tag in an Indian apparel store is GST-inclusive, so the counter works
backwards: the customer pays the tag less the offer, and the tax already sits
inside that number. Two things follow, and both are easy to get subtly wrong.

**Which slab.** `CONTEXT.md` locks it: 5% at or under ₹2,500 per piece, 18% above,
decided on the GST-**exclusive**, post-discount, per-piece price. That is
circular on the face of it - the base depends on the rate and the rate depends on
the base - and it resolves cleanly because the mapping is monotone: price the
piece at the lower rate first, and if the base that produces is still inside the
threshold, the lower rate is the right one. There is no tie to break and the MRP
can never cross the boundary its own slab was chosen by. `ptmapper.pricing`
anchors on the same base going the other way, which is what keeps a piece priced
by head office and the same piece sold at the counter in the same slab.

**Where the half-paisa goes.** The base is rounded half-up and the tax is the
remainder, never rounded separately - so base + tax is exactly what the customer
paid, on every line, with no drift to explain at the end of the day.

Both functions are advisory here: the till has already priced the bill and
printed it. The server recomputes to *compare* (contract step 12), and a
disagreement raises a flag rather than a refusal.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from decimal import ROUND_HALF_UP, Decimal

_HUNDRED = Decimal("100")


@dataclass(frozen=True)
class Slab:
    """One date-effective apparel GST slab, in the units the till speaks."""

    threshold_paise: int
    rate_below: Decimal
    rate_above: Decimal
    effective_from: date

    def rate_for_base(self, base_paise: int) -> Decimal:
        return self.rate_below if base_paise <= self.threshold_paise else self.rate_above


@dataclass(frozen=True)
class TaxSplit:
    """A GST-inclusive line, taken apart."""

    rate: Decimal
    base_paise: int
    gst_paise: int


def base_from_inclusive(inclusive_paise: int, rate: Decimal) -> int:
    """The GST-exclusive base inside a tax-inclusive amount, rounded half-up."""
    return int(
        (Decimal(inclusive_paise) * _HUNDRED / (_HUNDRED + rate)).quantize(
            Decimal(1), rounding=ROUND_HALF_UP
        )
    )


def split_inclusive(inclusive_paise: int, rate: Decimal) -> TaxSplit:
    """Split a tax-inclusive amount at a known rate. Tax is the remainder."""
    base = base_from_inclusive(inclusive_paise, rate)
    return TaxSplit(rate=rate, base_paise=base, gst_paise=inclusive_paise - base)


def split_line(inclusive_paise: int, qty: int, slab: Slab) -> TaxSplit:
    """Split a line, choosing the slab from its per-piece GST-exclusive price.

    `qty` matters because the threshold is per piece: two ₹2,000 shirts on one
    line are two 5% pieces, not one 18% line.
    """
    if qty <= 0:  # pragma: no cover - the model's CHECK refuses this first
        raise ValueError("a line must carry a positive quantity to be priced")
    per_piece = Decimal(inclusive_paise) / qty
    base_at_low = int(
        (per_piece * _HUNDRED / (_HUNDRED + slab.rate_below)).quantize(
            Decimal(1), rounding=ROUND_HALF_UP
        )
    )
    rate = slab.rate_for_base(base_at_low)
    return split_inclusive(inclusive_paise, rate)

File: app/frontend/src/till/pricing.ts
// MRP-inclusive tax arithmetic at the counter (#180, D10 step 3).
//
// The TypeScript mirror of `sell/pricing.py`, and it exists because the till
// prices a scan with no network: the price tag is GST-inclusive, so the counter
// works backwards from what the customer pays to the tax already sitting inside
// it.
//
// Two rules, both from `CONTEXT.md`, both easy to get subtly wrong:
//
// **Which slab.** 5% at or under ₹2,500 per piece and 18% above, decided on the
// GST-*exclusive*, post-discount, per-piece price. That looks circular - the base
// depends on the rate and the rate on the base - and resolves because the mapping
// is monotone: price the piece at the lower rate first, and if the base that
// produces is still inside the threshold, the lower rate is the right one.
//
// **Where the half-paisa goes.** The base is rounded half-up and the tax is the
// remainder, never rounded on its own, so base + tax is exactly what the customer
// handed over on every line.
//
// All of it in integers. A rate arrives as a two-decimal string ("5.00") and is
// carried as hundredths of a percent, so nothing here ever touches a float - a
// bill that came out a paise different from the server's recomputation would
// raise a `gst_mismatch` flag every single time.

import type { TillGstSlab } from "./types";

/** ₹2,500 a piece, as the fallback when no slab has reached the device yet. */
const FALLBACK_SLAB: TillGstSlab = {
  hsn_prefix: "",
  threshold_paise: 250000,
  rate_below: "5.00",
  rate_above: "18.00",
  effective_from: "2025-09-22",
};

export interface TaxSplit {
  /** The rate applied, in the two-decimal form the bill and the server quote. */
  rate: string;
  base_paise: number;
  gst_paise: number;
}

/** A two-decimal percentage as hundredths of a percent: "5.00" → 500. */
export function rateHundredths(rate: string): number {
  const [whole, fraction = ""] = rate.trim().split(".");
  return Number(whole) * 100 + Number((fraction + "00").slice(0, 2));
}

/** The GST-exclusive base inside a tax-inclusive amount, rounded half-up.
 *
 *  Integer arithmetic throughout: `x * 100 / (100 + rate)` in floating point is
 *  a paise adrift often enough to matter across a day's bills. */
export function baseFromInclusive(inclusivePaise: number, rate: string): number {
  const denominator = 10_000 + rateHundredths(rate);
  const numerator = inclusivePaise * 10_000;
  return Math.floor((2 * numerator + denominator) / (2 * denominator));
}

/** Split a tax-inclusive amount at a known rate. The tax is the remainder. */
export function splitInclusive(inclusivePaise: number, rate: string): TaxSplit {
  const base = baseFromInclusive(inclusivePaise, rate);
  return { rate, base_paise: base, gst_paise: inclusivePaise - base };
}

/**
 * Split a whole line, choosing the slab from its per-piece exclusive price.
 *
 * `qty` matters because the threshold is per piece: two ₹2,000 shirts on one line
 * are two 5% pieces, not one 18% line.
 */
export function splitLine(inclusivePaise: number, qty: number, slab: TillGstSlab): TaxSplit {
  if (qty <= 0) throw new Error("a line must carry a positive quantity to be priced");
  // The per-piece base in one step. Dividing by `qty` first and rounding, then
  // taking the tax out of that, rounds twice - and the second rounding can push a
  // piece across the ₹2,500 boundary the server would have left below it.
  const denominator = qty * (10_000 + rateHundredths(slab.rate_below));
  const baseAtLow = Math.floor((2 * inclusivePaise * 10_000 + denominator) / (2 * denominator));
  const rate = baseAtLow <= slab.threshold_paise ? slab.rate_below : slab.rate_above;
  return splitInclusive(inclusivePaise, rate);
}

/**
 * The slab in force on `when`, preferring the row written for this HSN.
 *
 * Date-effective by construction (Rule 11): a bill is taxed by the slab that was
 * live on the day it was billed, and the dataset deliberately ships slabs whose
 * date has not arrived so an offline counter reaches October's rate in October.
 *
 * Where this and the server could disagree: two slabs sharing an
 * `effective_from` and both matching the HSN. The server's ordering does not
 * break that tie, so this takes the longer prefix - the more specific rule - and
 * the daily applied-versus-rulebook check would catch it if head office ever
 * wrote such a pair.
 */
export function slabFor(slabs: TillGstSlab[], hsn: string, when: string): TillGstSlab {
  const live = slabs
    .filter((s) => s.effective_from <= when)
    .sort(
      (a, b) =>
        b.effective_from.localeCompare(a.effective_from) ||
        b.hsn_prefix.length - a.hsn_prefix.length,
    );
  return (
    live.find((s) => s.hsn_prefix && hsn && hsn.startsWith(s.hsn_prefix)) ??
    live.find((s) => !s.hsn_prefix) ??
    live[0] ??
    FALLBACK_SLAB
  );
}
